src/Abp.Zero.EntityFramework/EntityHistory/EntityHistoryHelper.cs

```diff
@@ -487,54 +478,100 @@ private void UpdateChangeSet(DbContext context, EntityChangeSet changeSet)
+                    var propertyChange = entityChange.PropertyChanges.FirstOrDefault(pc => pc.PropertyName == propertyName);
+
+                    // TODO: fix new value comparison before truncation
+                    var newValue = propertyEntry.GetNewValue()?.ToJsonString().TruncateWithPostfix(EntityPropertyChange.MaxValueLength);
+
+                    if (propertyChange == null)
                     {
-                        var propertyEntry = entityEntry.Property(property.Name);
-                        var propertyChange = entityChange.PropertyChanges.FirstOrDefault(pc => pc.PropertyName == property.Name);
+                        if (propertyEntry.HasChanged())
+                        {
+                            var oldValue = propertyEntry.GetOriginalValue()?.ToJsonString().TruncateWithPostfix(EntityPropertyChange.MaxValueLength);
+
+                            // Add foreign key
+                            entityChange.PropertyChanges.Add(CreateEntityPropertyChange(oldValue, newValue, propertyInfo));
+                        }
 
-                        //make sure test case cover post saving (for foreign key update)
-                        if (propertyChange == null)
+                        continue;
+                    }
+
+                    if (propertyChange.OriginalValue == propertyChange.NewValue)
+                    {
+                        if (propertyChange.NewValue == newValue)
+                        {
+                            // No change
+                            entityChange.PropertyChanges.Remove(propertyChange);
+                        }
+                        else
                         {
-                            if (propertyEntry.HasChanged())
-                            {
-                                var propertyInfo = entityEntry.GetPropertyInfo(property.Name);
-
-                                // Add foreign key
-                                entityChange.PropertyChanges.Add(new EntityPropertyChange
-                                {
-                                    NewValue = propertyEntry.CurrentValue.ToJsonString(),
-                                    OriginalValue = propertyEntry.OriginalValue.ToJsonString(),
-                                    PropertyName = property.Name,
-                                    PropertyTypeFullName = propertyInfo.PropertyType.FullName
-                                });
-                            }
-
-                            continue;
+                            // Update foreign key
+                            propertyChange.NewValue = newValue;
                         }
+                    }
+                }
+
+                /* Update tracked properties */
+                var auditedPropertyChanges = entityChange.PropertyChanges
+                                             .Where(pc => !foreignKeyPropertyNames.Contains(pc.PropertyName))
+                                             .Distinct()
```

Comment:
Should not need `.Distinct()` as `entityChange.PropertyChanges` should already contain unique entries.

Reply:
👍